modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/DynamicEnableIndexingConcurrentSelfTest.java

```diff
@@ -0,0 +1,629 @@
+        IgniteEx srv1 = ignitionStart(serverConfiguration(1));
+        Ignite srv2 = ignitionStart(serverConfiguration(2));
+        srv1.cluster().state(ClusterState.ACTIVE);
+
+        createCache(srv1);
+        loadData(srv1, 0, LARGE_NUM_ENTRIES);
+
+        // Start index operation in blocked state.
+        CountDownLatch idxLatch1 = blockIndexing(srv1);
+        CountDownLatch idxLatch2 = blockIndexing(srv2);
+
+        IgniteInternalFuture<?> tblFut = enableIndexing(srv1);
+
+        U.await(idxLatch1);
+        U.await(idxLatch2);
+
+        // Start two more nodes and unblock index operation in the middle.
+        ignitionStart(serverConfiguration(3));
+
+        unblockIndexing(srv1);
+        unblockIndexing(srv2);
+
+        ignitionStart(serverConfiguration(4));
+
+        awaitPartitionMapExchange();
+
+        tblFut.get();
+
+        for (Ignite g: G.allGrids()) {
+            assertEquals(LARGE_NUM_ENTRIES, query(g, SELECT_ALL_QUERY).size());
+
+            performQueryingIntegrityCheck(g);
+
+            checkQueryParallelism((IgniteEx)g, cacheMode);
+        }
+    }
+
+    /** Test concurrent put remove when enabling indexing. */
+    @Test
+    public void testConcurrentPutRemove() throws Exception {
+        CountDownLatch finishLatch = new CountDownLatch(4);
+
+        // Start several nodes.
+        IgniteEx srv1 = ignitionStart(serverConfiguration(1), finishLatch);
+        ignitionStart(serverConfiguration(2), finishLatch);
+        ignitionStart(serverConfiguration(3), finishLatch);
+        ignitionStart(serverConfiguration(4), finishLatch);
+
+        srv1.cluster().state(ClusterState.ACTIVE);
+
+        createCache(srv1);
+        loadData(srv1, 0, LARGE_NUM_ENTRIES);
+
+        // Start data change operations from several threads.
+        final AtomicBoolean stopped = new AtomicBoolean();
+
+        IgniteInternalFuture<?> task = multithreadedAsync(() -> {
+            while (!stopped.get()) {
+                Ignite node = grid(ThreadLocalRandom.current().nextInt(1, 5));
+
+                ThreadLocalRandom rnd = ThreadLocalRandom.current();
+
+                int i = rnd.nextInt(0, LARGE_NUM_ENTRIES);
+
+                BinaryObject val = node.binary().builder(POI_CLASS_NAME)
+                    .setField(NAME_FIELD_NAME, "POI_" + i, String.class)
+                    .setField(LATITUDE_FIELD_NAME, rnd.nextDouble(), Double.class)
+                    .setField(LONGITUDE_FIELD_NAME, rnd.nextDouble(), Double.class)
+                    .build();
+
+                IgniteCache<Object, BinaryObject> cache = node.cache(POI_CACHE_NAME).withKeepBinary();
+
+                try {
+                    if (ThreadLocalRandom.current().nextBoolean())
+                        cache.put(i, val);
+                    else
+                        cache.remove(i);
+                }
+                catch (CacheException e) {
+                    if (!X.hasCause(e, TransactionSerializationException.class))
```

Comment:
Why we ignore `TransactionSerializationException`?

Reply:
Because it is normal situation, see documentation to TRANSACTIONAL_SHAPSHOT